system_modules/napsequence/src/sequenceplayer.cpp

```diff
@@ -316,96 +329,100 @@ namespace nap
 
     double SequencePlayer::getDuration() const
     {
-        return mSequence->mDuration;
+        return mDuration;
     }
 
 
     void SequencePlayer::setPlayerTime(double time)
     {
         {
-            std::lock_guard<std::mutex> lock(mMutex);
-            mTime = math::clamp<double>(time, 0.0, mSequence->mDuration);
+            std::lock_guard l(mTimeMutex);
+            mTime.store(math::clamp<double>(time, 0.0, getDuration()));
         }
-        playerTimeChanged(*this, mTime);
+        playerTimeChanged(*this, mTime.load());
     }
 
 
     void SequencePlayer::setPlaybackSpeed(float speed)
     {
-        {
-            std::lock_guard<std::mutex> lock(mMutex);
-            mSpeed = speed;
-        }
+        mSpeed = speed;
         playbackSpeedChanged(*this, speed);
     }
 
 
     double SequencePlayer::getPlayerTime() const
     {
-        return mTime;
+        return mTime.load();
     }
 
 
     bool SequencePlayer::getIsPlaying() const
     {
-        return mIsPlaying;
+        return mIsPlaying.load();
     }
 
 
     bool SequencePlayer::getIsPaused() const
     {
-        return mIsPaused;
+        return mIsPaused.load();
     }
 
 
     void SequencePlayer::setIsLooping(bool isLooping)
     {
-        std::lock_guard<std::mutex> lock(mMutex);
-        mIsLooping = isLooping;
+        mIsLooping.store(isLooping);
     }
 
 
     bool SequencePlayer::getIsLooping() const
     {
-        return mIsLooping;
+        return mIsLooping.load();
     }
 
 
     float SequencePlayer::getPlaybackSpeed() const
     {
-        return mSpeed;
+        return mSpeed.load();
     }
 
 
+    void SequencePlayer::advanceTime(double deltaTime)
+    {
+        std::lock_guard l(mTimeMutex);
+        double time = mTime.load();
+        time += deltaTime * static_cast<double>(mSpeed.load());
+        if(mIsLooping.load())
+        {
+            if(time < 0.0)
+            {
+                time = getDuration() + time;
+            }else if(time > getDuration())
+            {
+                time = fmod(time, getDuration());
+            }
+        } else
+        {
+            time = math::clamp<double>(time, 0.0, getDuration());
+        }
+        mTime.store(time);
+    }
+
+
+
     void SequencePlayer::tick(double deltaTime)
     {
-        if(mIsPlaying)
+        if(mIsPlaying.load())
         {
-            // notify lister, so data model of sequence and data of player can be modified by listeners to this signal
+            // notify listener, so data model of sequence and data of player can be modified by listeners to this signal
             preTick.trigger(*this);
 
-            // Update time and adapters thread safe
+            // Update time
+            if(!mIsPaused.load())
```

Comment:
For future reference, you don't have to explicitly load, the operator is overloaded